### PR TITLE
Add support for Incident Custom Fields and sanitize Orchestration PUT request

### PR DESCRIPTION
**Pull request by swbradshaw**
This PR adds support for Incident Custom Fields on the Event Orchestration path. This feature will be released in January 2024. In addition it fixes the API call for updating the Event Orchestration.

Changes:

- Added new `EventOrchestrationPathIncidentCustomFieldUpdate` struct which is an ID/Value pair of the custom field
- Added this new property to existing `EventOrchestrationPathRuleActions` struct
- Updated test in `pagerduty/event_orchestration_path_test.go` for new action type
- In the process of testing Getting/Saving a Service Orchestration, I determined there was a bug in this client where it was sending `null` values for arrays. This is not allowed by the API and results in an HTTP 400 Bad Request. I've added functions `sanitizeOrchestrationPath` and `sanitizeActions`  to replace null values with empty arrays. 

The following is a test client I wrote to test this library. The code below simply gets the service orchestration and then updates it with the same object from the get request. Without the sanitize change, this library throws an error.

```
func main() {
	client, err := pagerduty.NewClient(&pagerduty.Config{
		Token: os.Getenv("PAGERDUTY_TOKEN")})
	if err != nil {
		panic(err)
	}

	servicesResp, _, err := client.Services.List(&pagerduty.ListServicesOptions{})
	if err != nil {
		panic(err)
	}
	serviceId := servicesResp.Services[0].ID
	
	servicePathResp, _, err := client.EventOrchestrationPaths.Get(serviceId, "service")
	if err != nil {
		panic(err)
	}
	
	updatePayload, _, err := client.EventOrchestrationPaths.Update(serviceId, "service", servicePathResp)
	if err != nil {
		panic(err)
	}

}

```

### CODE REVIEW

File: pagerduty/event_orchestration_path_test.go [hunk 1]
```diff
@@ -601,14 +627,14 @@ func TestEventOrchestrationPathUnroutedUpdate(t *testing.T) {
 			},
 		},
 		Warnings: []*EventOrchestrationPathWarning{
-			&EventOrchestrationPathWarning{
```
Comment (swbradshaw) [inline pagerduty/event_orchestration_path_test.go:604]: Annotate: `redundant type from array, slice, or map composite literal - simplifycompositelit(default)`

File: pagerduty/event_orchestration_path.go [hunk 2]
```diff
@@ -170,6 +176,7 @@ func (s *EventOrchestrationPathService) Update(id string, pathType string, orche
 func (s *EventOrchestrationPathService) UpdateContext(ctx context.Context, id string, pathType string, orchestrationPath *EventOrchestrationPath) (*EventOrchestrationPathPayload, *Response, error) {
 	u := orchestrationPathUrlBuilder(id, pathType)
 	v := new(EventOrchestrationPathPayload)
+	sanitizeOrchestrationPath(orchestrationPath)
```
Comment (alenapan) [inline pagerduty/event_orchestration_path.go]: Thank you for fixing this! Could you also add a test verifying that if you pass all of these properties as `nil` the payload is sent to the API with empty arrays instead? The test can be added for any orchestration type but the real world scenario would be a service orchestration since it supports all of these actions.

Comment (swbradshaw) [reply pagerduty/event_orchestration_path.go]: Thanks. I just added some tests around this and had to address a previous test.

Comment (alisonwarner-pd) [reply pagerduty/event_orchestration_path.go]: ty!
My go is a little rusty rusty but I'll give a lgtm because nothing is blocking

**Review by alenapan — APPROVED** [review]

File: pagerduty/event_orchestration_path.go (hunk 2 above)
Comment (imjaroiswebdev) [reply pagerduty/event_orchestration_path.go]: Hey @swbradshaw thank you for introducing this fix and because it is a patch, I'd prefer to move this patch to another PR in the current one can depend on before it gets merged. On top of that, being that this PR is not planned to be merged soon, because it relies on having its corresponding API Reference updated too, then We would be able to ship this bugfix first. So, could you please move this fix to another PR?

File: pagerduty/event_orchestration_path_test.go [hunk 3]
```diff
@@ -250,6 +250,55 @@ func TestEventOrchestrationPathGetServiceActiveStatus(t *testing.T) {
...
+		},
+	}
+
+	var url = fmt.Sprintf("%s/E-ORC-1/global", eventOrchestrationBaseUrl)
+
+	mux.HandleFunc(url, func(w http.ResponseWriter, r *http.Request) {
+		testMethod(t, r, "PUT")
+		v := new(EventOrchestrationPathPayload)
+		json.NewDecoder(r.Body).Decode(v)
+		if !reflect.DeepEqual(v.OrchestrationPath, input) {
```
Comment (imjaroiswebdev) [inline pagerduty/event_orchestration_path_test.go]: Test to covert the patch from a regression should be happening at this point, same as You did bellow in `TestEventOrchestrationPathGlobalUpdate`, where You validate the following assertion...

```go
		if !reflect.DeepEqual(string(actionJson), "[]") {
			t.Errorf("empty action array should be []")
		}
```

Because what the test should check is that the input payload is now sanitized and as a result will no longer produce  a **Bad Request** error output from API.

File: pagerduty/event_orchestration_path_test.go [hunk 4]
```diff
@@ -463,6 +542,9 @@ func TestEventOrchestrationPathGlobalUpdate(t *testing.T) {
 		},
 	}
 
+	// Before comparison, we sanitize the response to make it match the already sanitized request
+	sanitizeOrchestrationPath(resp.OrchestrationPath)
```
Comment (imjaroiswebdev) [inline pagerduty/event_orchestration_path_test.go]: 1. I suggest you to please test `sanitizeOrchestrationPath` helper function as a unit, instead of introducing adaptation of the response in the middle of the test.

2. In this case the comparison mocked data should resemble the expected output data, instead of changing the output of `client.EventOrchestrationPaths.Update` to make it equal to the mocked `want` expected output, it's ok and the result is the same, however the order in this must happen is backwards. Thus, I ask you to please make the necessary updates @swbradshaw

**Review by imjaroiswebdev — CHANGES_REQUESTED** [review]
Thank you for adding this very needed fix @swbradshaw, however before moving on, there are a few comments which need to be addressed to continue with the merge. So, please if You find We need to discuss this further don't hesitate in reaching me out. ✌🏽

File: pagerduty/event_orchestration_path_test.go (hunk 4 above)
Comment (swbradshaw) [reply pagerduty/event_orchestration_path_test.go]: > In this case the comparison mocked data should resemble the expected output data

The `input` that goes into the `client.EventOrchestrationPaths.Update` is sanitized, and then used in the `want` variable.  Thus, the `want` variable is sanitized and the response is not.  The test will never pass anymore without this line. 

Ultimately the problem is this library is incomplete for Event Orchestration. In addition to sanitizing the request, it should also sanitized the response. Without this, there is extra code needed to fix the response (Example, the [TF code has all these flattenEventOrchestrationXXXXX functions](https://github.com/PagerDuty/terraform-provider-pagerduty/blob/master/pagerduty/event_orchestration_path_util.go) which should not be needed.)

File: pagerduty/event_orchestration_path.go (hunk 2 above)
Comment (swbradshaw) [reply pagerduty/event_orchestration_path.go]: I don't have time to create a new PR for this. Given this library can't create orchestrations without supplemental code and no one has complained, I don't think a patch needs to be merged sooner than this.  If you don't want it here, I'll just remove all these changes related to sanitizing the request.

File: pagerduty/event_orchestration_path_test.go (hunk 3 above)
Comment (swbradshaw) [reply pagerduty/event_orchestration_path_test.go]: This check is already happening in line 294 below, as the `input` that goes into `client.EventOrchestrationPaths.Update` is sanitized. I can move this up if you want though?

**Review by imjaroiswebdev — APPROVED** [review]
LGTM! 👍🏽
